File: src/main/java/cope/nebula/client/feature/module/combat/AutoTotem.java
package cope.nebula.client.feature.module.combat;

import cope.nebula.client.feature.module.Module;
import cope.nebula.client.feature.module.ModuleCategory;
import cope.nebula.client.value.Value;
import cope.nebula.util.world.entity.EntityUtil;
import cope.nebula.util.world.entity.player.inventory.InventorySpace;
import cope.nebula.util.world.entity.player.inventory.InventoryUtil;
import cope.nebula.util.world.entity.player.inventory.task.InventoryTask;
import cope.nebula.util.world.entity.player.inventory.task.InventoryTaskHandler;
import net.minecraft.init.Items;
import net.minecraft.inventory.ClickType;
import net.minecraft.item.Item;
import net.minecraft.item.ItemSword;
import net.minecraft.util.EnumHand;

public class AutoTotem extends Module {
    public AutoTotem() {
        super("AutoTotem", ModuleCategory.COMBAT, "Manages your offhand");
    }

    public static final Value<Mode> mode = new Value<>("Mode", Mode.CRYSTAL);
    public static final Value<Float> health = new Value<>("Health", 14.0f, 1.0f, 20.0f);
    public static final Value<Integer> delay = new Value<>("Delay", 2, 0, 10);
    public static final Value<Boolean> await = new Value<>("Await", false);
    public static final Value<Boolean> gap = new Value<>("Gap", true);

    private final InventoryTaskHandler taskHandler = new InventoryTaskHandler();
    private Item item;

    @Override
    protected void onActivated() {
        EVENT_BUS.register(taskHandler);
    }

    @Override
    protected void onDeactivated() {
        EVENT_BUS.unregister(taskHandler);
        taskHandler.clean();
    }

    @Override
    public void onTick() {
        if (!taskHandler.execute(delay.getValue() * 50L, await.getValue())) {
            return;
        } else {
            item = null;
        }

        if (willDieUponFall() || EntityUtil.getHealth(mc.player) <= health.getValue()) {
            item = Items.TOTEM_OF_UNDYING;
        } else {
            if (InventoryUtil.getHeld(EnumHand.MAIN_HAND).getItem() instanceof ItemSword &&
                    gap.getValue() &&
                    mc.gameSettings.keyBindUseItem.isKeyDown()) {
                item = Items.GOLDEN_APPLE;
            } else {
                item = mode.getValue().item;
            }
        }

        if (InventoryUtil.getHeld(EnumHand.OFF_HAND).getItem().equals(item)) {
            return;
        }

        int slot = InventoryUtil.getSlot(InventorySpace.BOTH, (s) -> !s.isEmpty() && s.getItem().equals(item));
        if (slot != -1) {
            int packetSlotId = InventoryUtil.normalize(slot);

            taskHandler.addTask(new InventoryTask(packetSlotId, ClickType.PICKUP));
            taskHandler.addTask(new InventoryTask(InventoryUtil.OFFHAND_SLOT, ClickType.PICKUP));
            if (!InventoryUtil.getHeld(EnumHand.OFF_HAND).isEmpty()) {
                taskHandler.addTask(new InventoryTask(packetSlotId, ClickType.PICKUP));
            }
        }
    }

    /**
     * @return if given the fall damage, we will die
     */
    private boolean willDieUponFall() {
        return ((mc.player.fallDistance - 3.0f) / 2.0f) + 3.5f >= EntityUtil.getHealth(mc.player);
    }

    public enum Mode {
        TOTEM(Items.TOTEM_OF_UNDYING),
        GAPPLE(Items.GOLDEN_APPLE),
        CRYSTAL(Items.END_CRYSTAL),
        EXP(Items.EXPERIENCE_BOTTLE);

        private final Item item;

        Mode(Item item) {
            this.item = item;
        }
    }
}

File: src/main/kotlin/wtf/nebula/client/feature/module/combat/AutoTotem.kt
package wtf.nebula.client.feature.module.combat

import me.bush.eventbuskotlin.EventListener
import me.bush.eventbuskotlin.listener
import net.minecraft.init.Items
import net.minecraft.item.ItemSword
import wtf.nebula.client.Nebula
import wtf.nebula.client.event.tick.TickEvent
import wtf.nebula.client.feature.module.Module
import wtf.nebula.client.feature.module.ModuleCategory
import wtf.nebula.client.manager.Task
import wtf.nebula.client.manager.TaskList
import wtf.nebula.util.inventory.InventoryUtil

class AutoTotem : Module(ModuleCategory.COMBAT, "Automatically places totems in your offhand") {
    val item by setting("Item", Item.CRYSTAL)
    val delay by int("Delay", 2, 0..10)
    val swordGap by bool("Sword GApple", true)
    val hotbar by bool("Hotbar", true)
    val health by float("Health", 14.0f, 2.0f..20.0f)

    private var currentItem = Items.TOTEM_OF_UNDYING

    override fun getDisplayInfo(): String? {
        val offhandStack = mc.player.heldItemOffhand
        if (offhandStack.isEmpty) {
            return null
        }

        return mc.player.inventory.mainInventory
            .filter { it.item == offhandStack.item }
            .sumOf { it.count }
            .toString()
    }

    override fun onDeactivated() {
        super.onDeactivated()
        Nebula.inventoryManager.removeAllTasks(hashCode())
    }

    @EventListener
    private val tickListener = listener<TickEvent> {
        if (Nebula.inventoryManager.isEmpty(hashCode())) {
            if (canDieFromFallDamage() || mc.player.health < health) {
                currentItem = Items.TOTEM_OF_UNDYING
            }

            else {
                currentItem = item.item
                if (swordGap && mc.gameSettings.keyBindUseItem.isKeyDown && mc.player.heldItemMainhand.item is ItemSword) {
                    currentItem = Items.GOLDEN_APPLE
                }
            }

            handleItems()
        }
    }

    private fun handleItems() {
        if (mc.player.heldItemOffhand.item == currentItem) {
            return
        }

        if (hotbar && Nebula.inventoryManager.getHeldItemStack().item == currentItem) {
            return
        }

        var slot = -1
        for (i in 0 until 36) {
            val stack = mc.player.inventory.getStackInSlot(i)
            if (!stack.isEmpty && stack.item == currentItem) {
                slot = i
            }
        }

        if (slot == -1) {
            return
        }

        val serverSlot = if (slot < 9) slot + 36 else slot
        val floatingItem = !mc.player.heldItemOffhand.isEmpty

        var taskList = TaskList(hashCode())

        taskList += Task(slot = serverSlot, delay = delay)
        taskList += Task(slot = InventoryUtil.OFFHAND_SLOT, delay = delay)

        if (floatingItem) {
            taskList += Task(slot = serverSlot, delay = delay)
        }

        Nebula.inventoryManager.submitTasks(taskList)
    }

    private fun canDieFromFallDamage(): Boolean {
        return ((mc.player.fallDistance - 3.0f) / 2.0f) + 3.5f >= mc.player.health + mc.player.absorptionAmount
    }

    enum class Item(val item: net.minecraft.item.Item) {
        TOTEM(Items.TOTEM_OF_UNDYING),
        CRYSTAL(Items.END_CRYSTAL),
        GAPPLE(Items.GOLDEN_APPLE),
        EXP(Items.EXPERIENCE_BOTTLE)
    }
}